bot/bot.go

```diff
@@ -0,0 +1,224 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/bwmarrin/discordgo"
+)
+
+var botID string
+
+// TODO: Turn this into a concurrent set to handle safe
+// discord server removals
+var subscribedServers map[string]string
+
+func Run() {
+	subscribedServers = make(map[string]string)
+
+	session, err := discordgo.New("Bot " + config.DiscordBotToken)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	user, err := session.User("@me")
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	botID = user.ID
+	session.AddHandler(onJoinGuild)
+
+	err = session.Open()
+	if err != nil {
+		return
+	}
+	defer session.Close()
+
+	// This Discord client doesn't have a way to attach event handlers to the
+	// heartbeat, therefore we need our own timer to post to all the servers.
+	ticker := time.NewTicker(time.Duration(config.TickRateInSeconds) * time.Second)
+	quit := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				// Loop over Discord servers and do a thing
+				for guildID, _ := range subscribedServers {
+					// TODO: Run these in parallel?
+					postNextEvent(session, guildID)
+				}
+			case <-quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// In order to create/post events per discord server, we need to know which
+// server we're currently subscribed to.
+func onJoinGuild(session *discordgo.Session, event *discordgo.GuildCreate) {
+	// TODO: We need a way to know which channel to post to per discord server
+	// So we'll do that assignment here and then know on lookup of server
+	// on post to send the post
+	subscribedServers[event.ID] = "<channel id>"
+
+	postNextEvent(session, event.ID)
+}
+
+// TODO: We need a handler for when the bot leaves a guild so we can remove it from
+// the guild from the subscribedGuilds list. There doesn't seem to be a GuildRemove event
+// on first glance. Will need to setup a second Discord server to test the multi-server setup.
+
+func postNextEvent(session *discordgo.Session, guildID string) {
+
+	if !shouldGetNextMeetupEvent(guildID) {
+		return
+	}
+
+	// TODO: Figure out some caching mechanism so we only fetch this once
+	// for all servers that need updating
+	meetupEvent := getNextMeetupEvent()
+
+	// Use: session.GuildScheduledEvents()
+	// To grab current events in discord, and see if our next fetched event is
+	// already there.
+	// If not, create the event in Discord and post the meetup event in channel
+	// If yes, don't post to channel -- this is our final check to not blast
+	// servers upon joining them (helps for the case when the bot goes down in
+	// and boots back up in the middle of a cycle)
+
+	channelID := subscribedServers[guildID]
+	fmt.Printf("[%s] Posting next event to %s: %s\n", guildID, channelID, meetupEvent)
+	// session.ChannelMessageSend(channelID, "some message about the upcoming event")
+}
+
+func shouldGetNextMeetupEvent(guildID string) bool {
+	now := time.Now()
+	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
```

Comment:
Need to switch this to at least server time or add configuration option because server time may not be for the timezone we want this blasting at 10:00am.